### PR TITLE
781 - collections consume invitation

### PR DESCRIPTION
**Pull request by kkoehn**
Currently a user can use an invitation as many times as they want, which does not feel intuitive.
This PR also adds an overview of how many open invitations a collection has

depends on #1130

### CODE REVIEW

File: app/controllers/collections_controller.rb [hunk 1]
```diff
@@ -95,14 +97,15 @@ def view_shared
   end
 
   def save_shared
-    return redirect_to user_messages_path(current_user), alert: t('.errors.already_member') if @collection.users.include? current_user
+    user = current_user
```
Comment (Dome-GER) [inline app/controllers/collections_controller.rb:100]: Why is the `current_user` assigned here? 🤔

File: app/controllers/collections_controller.rb [hunk 2]
```diff
@@ -95,14 +97,15 @@ def view_shared
...
-      redirect_to @collection, notice: t('.success_notice')
-    else
-      redirect_to users_messages_path, alert: t('common.errors.something_went_wrong')
-    end
+    @collection.users << user
+    message = Message.received_by(user).find_by(param_type: 'collection', param_id: @collection.id)
+    message.mark_as_deleted(user)
+    message.save
+
+    redirect_to @collection, notice: t('.success_notice')
```
Comment (Dome-GER) [inline app/controllers/collections_controller.rb:108]: So the success is defined now? Saving the message can still fail, no?

File: app/views/collections/show.html.slim [hunk 3]
```diff
@@ -13,12 +13,19 @@
...
+      - if action_name != 'view_shared'
+        span
+          i.fa-solid.fa-users
+          - if @collection.users.size == 1
+            =< t('.no_other_user')
+          - else
+            =< t('.num_of_other_users', count: @collection.users.size - 1)
+        span
+          i.fa-regular.fa-envelope
+          - if @num_of_invites == 0
```
Comment (Dome-GER) [inline app/views/collections/show.html.slim:25]:
```suggestion
          - if @num_of_invites.zero?
```

I tend to prefer `zero?` (similar to `empty?`) for readability over `== 0` unless `0` is meant as some math value, i.e. coordinate or whatever.

**Review by Dome-GER — APPROVED** [review]
Looks good, merely minor suggestions that I consider non-blocking.

File: app/controllers/collections_controller.rb (hunk 1 above)
Comment (kkoehn) [reply app/controllers/collections_controller.rb:100]: We need `current_user` 5 times in this method and with that the Metrics/AbcSize-Cop triggers. An assignment beforehand alleviates that issue.

File: app/controllers/collections_controller.rb [hunk 4]
```diff
@@ -94,14 +96,19 @@ def view_shared
...
 
-    @collection.users << current_user
-    if @collection.save
-      redirect_to @collection, notice: t('.success_notice')
+    ActiveRecord::Base.transaction(requires_new: true) do
+      @collection.users << current_user
+      message = Message.received_by(current_user).find_by(param_type: 'collection', param_id: @collection.id)
+      message.mark_as_deleted(current_user)
+      message.save!
+    rescue ActiveRecord::ActiveRecordError
```
Comment (Dome-GER) [inline app/controllers/collections_controller.rb:107]: I tend to be more specific and rather list the errors I'd expect here. If something else is raised, we will notice in the monitoring.

File: spec/controllers/collections_controller_spec.rb [hunk 5]
```diff
@@ -78,6 +83,19 @@
       expect(response.body).to include(I18n.t('collections.show.no_other_user'))
     end
 
+    context 'when there are pending invites' do
+      before do
+        2.times do
+          create(:message, sender: user, recipient: create(:user), param_type: 'collection', param_id: collection.id, text: 'Invitation')
```
Comment (Dome-GER) [inline spec/controllers/collections_controller_spec.rb:89]: With factory_bot, `create_list` would be the built-in way to achieve this ;)

**Review by Dome-GER — APPROVED** [review]

File: app/controllers/collections_controller.rb (hunk 4 above)
Comment (kkoehn) [reply app/controllers/collections_controller.rb:107]: I don't really expect any errors here. I just want to make sure, that the database is in a consistent state. The user-feedback was just a bonus. As far as I see, any error is unexpected here and should end up in the monitoring. Should I just remove the `rescue` then?

Comment (Dome-GER) [reply app/controllers/collections_controller.rb:107]: I'd say so.

File: spec/controllers/collections_controller_spec.rb (hunk 5 above)
Comment (kkoehn) [reply spec/controllers/collections_controller_spec.rb:89]: Is there a way to make sure that `create(:user)` is called each message individually?

Comment (Dome-GER) [reply spec/controllers/collections_controller_spec.rb:89]: Sorry, I don't get the question. What's the purpose?

Comment (kkoehn) [reply spec/controllers/collections_controller_spec.rb:89]: I want to create two messages, and we have a uniqueness validation for the message if it is an invitation, ensuring that each recipient can only receive one invitation per collection:
```ruby
validates :param_id, uniqueness: {scope: %i[recipient_id param_type]}, if: -> { param_type == 'collection' }
```
Therefore each message would need its own user. If I replace the current `create` with a `create_list` the `create(:user)` is evaluated first and both messages have the same user, which fails the validation.

File: app/controllers/collections_controller.rb (hunk 2 above)
Comment (kkoehn) [reply app/controllers/collections_controller.rb]: Added a transaction to ensure database-consistency

File: app/controllers/collections_controller.rb (hunk 4 above)
Comment (kkoehn) [reply app/controllers/collections_controller.rb]: done

**Comment by MrSerth** [comment]
I just stumbled across the collection statistics, which I really like. Still, I have minor clarification questions: 

1. Currently, we show all four lines (visibility, member count, invitation count, favorite count) to anyone if the collection is public. I was wondering whether we should restrict some of the information (i.e., member count and/or pending invitation count) to members/admins only? What do you think, @kkoehn?

  Bildschirmfoto 2024-05-26 um 11 18 47

  ... and actually, I noticed that we hide this information for "view_shared" already:

  Bildschirmfoto 2024-05-26 um 11 25 05

2. The text "There are 2 pending invitations" does include **accepted** invitations as well, not only **pending** ones:
  
  Bildschirmfoto 2024-05-26 um 11 27 21

Can we fix these minor things, please? Sorry for the late review here 😇.

**Pull request merged**